error when run

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4e8cc6]

goroutine 132 [running]:
net.(*Resolver).lookupAddr(0x0, 0x590ec0, 0xc42001a088, 0xc42001aad0, 0xc, 0x4297e9, 0x8, 0xc420058080, 0x0, 0xc4200eeef0)
        /usr/lib/go-1.10/src/net/lookup_unix.go:189 +0x26
net.(*Resolver).LookupAddr(0x0, 0x590ec0, 0xc42001a088, 0xc42001aad0, 0xc, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/lib/go-1.10/src/net/lookup.go:390 +0x5a
main.worker(0xc42001aad0, 0xc, 0xc42001a2e0, 0xc420064120)
        /root/go/src/github.com/hakluke/hakrevdns/main.go:35 +0xca
created by main.main
        /root/go/src/github.com/hakluke/hakrevdns/main.go:57 +0x17e

```

--- a/README.md
+++ b/README.md
@@ -12,6 +12,18 @@ This can be a useful way of finding domains and subdomains belonging to a compan
 go get github.com/hakluke/hakrevdns
 ```
 
+### Alternative Installation using automated install script (Linux)
+
+#### Note: Requires ```wget``` in order to download the golang tarball
+
+  - Installs compatible version of golang
+  - Builds hakrevdns and places the binary in ```/usr/bin```
+  
+```sh
+chmod +x install.sh
+./install.sh
+```
+
 ## Usage
 Pipe a list of IP addresses into the tool, for example:
 
@@ -44,6 +56,7 @@ Application Options:
   -r, --resolver=          IP of the DNS resolver to use for lookups
   -P, --protocol=[tcp|udp] Protocol to use for lookups (default: udp)
   -p, --port=              Port to bother the specified DNS resolver on (default: 53)
+  -d, --domain             Output only domains
 
 Help Options:
   -h, --help               Show this help message
@@ -54,6 +67,17 @@ If you want to use a resolver not specified by you OS, say: 1.1.1.1, try this:
 ```sh
 hakluke~$ echo "173.0.84.110" | hakrevdns -r 1.1.1.1
 173.0.84.110    he.paypal.com.
+```
+
+If you wish to obtain only a list of domains, perhaps to sort out working ones you may use:
+
+```sh
+$ echo "173.0.84.110" | hakrevdns -d
+```
+For filtering out only working domains, you may additionally pipe it through [httprobe](https://github.com/tomnomnom/httprobe):
+
+```sh
+$ echo "173.0.84.110" | hakrevdns -d | httprobe
 ```
 
 ## Contributors

--- a/README.md
+++ b/README.md
@@ -9,7 +9,7 @@ This can be a useful way of finding domains and subdomains belonging to a compan
 ## Installation
 
 ```sh
-go get github.com/hackerpain/hakrevdns
+go get github.com/hakluke/hakrevdns
 ```
 
 ### Alternative Installation using automated install script (Linux)

--- a/README.md
+++ b/README.md
@@ -9,7 +9,7 @@ This can be a useful way of finding domains and subdomains belonging to a compan
 ## Installation
 
 ```sh
-go get github.com/hakluke/hakrevdns
+go get github.com/hackerpain/hakrevdns
 ```
 
 ### Alternative Installation using automated install script (Linux)

--- a/README.md
+++ b/README.md
@@ -17,7 +17,7 @@ go get github.com/hakluke/hakrevdns
 #### Note: Requires ```wget``` in order to download the golang tarball
 
   - Installs compatible version of golang
-  - Builds hakrevdns and places the binary in ```/usr/bin```
+  - Builds hakrevdns and places the binary in ```/usr/local/bin```
   
 ```sh
 chmod +x install.sh

--- a/install.sh
+++ b/install.sh
@@ -4,10 +4,13 @@
 
 if [ -z $GOPATH ]
 then
+  echo -e "\e[32m Installing golang \e[0m"
   wget https://dl.google.com/go/go1.14.src.tar.gz && tar -C /usr/local -xzf go1.14.src.tar.gz
   echo "export PATH=$PATH:/usr/local/go/bin:$GOPATH/bin" >> ~/.bashrc && source ~/.bashrc
+  echo -e "\e[32m Installed golang. Test it out by running \"go version\" \e[0m"
 fi
 
+echo -e "\e[32m Building hakrevdns \e[0m"
 go build && mv hakrevdns /usr/local
 
 if [ -e /usr/bin/hakrevdns ]

--- a/install.sh
+++ b/install.sh
@@ -1,0 +1,20 @@
+# Purpose of this file is to facilitate easy installation of golang on your debian box
+# Since most package managers have older versions of golang which may result in errors while running the tool
+# use this instead to install golang and the tool!
+
+if [ -z $GOPATH ]
+then
+  wget https://dl.google.com/go/go1.14.src.tar.gz && tar -C /usr/local -xzf go1.14.src.tar.gz
+  echo "export PATH=$PATH:/usr/local/go/bin:$GOPATH/bin" >> ~/.bashrc && source ~/.bashrc
+fi
+
+go build && mv hakrevdns /usr/local
+
+if [ -e /usr/bin/hakrevdns ]
+then
+  echo -e "\e[31mhakrevdns \e[32m has been successfully installed\e[0m"
+  echo -e "\e[32m Test it out by trying, echo "173.0.84.110" | hakrevdns \e[0m"
+fi
+
+
+# can't thank hakluke enough for making these amazing, fast tools, they give results in no time, amazingly easing out the workflow

--- a/install.sh
+++ b/install.sh
@@ -5,18 +5,18 @@
 if [ -z $GOPATH ]
 then
   echo -e "\e[32m Installing golang \e[0m"
-  wget https://dl.google.com/go/go1.14.src.tar.gz && tar -C /usr/local -xzf go1.14.src.tar.gz
+  wget https://dl.google.com/go/go1.14.src.tar.gz && sudo tar -C /usr/local -xzf go1.14.src.tar.gz
   echo "export PATH=$PATH:/usr/local/go/bin:$GOPATH/bin" >> ~/.bashrc && source ~/.bashrc
   echo -e "\e[32m Installed golang. Test it out by running \"go version\" \e[0m"
 fi
 
 echo -e "\e[32m Building hakrevdns \e[0m"
-go build && mv hakrevdns /usr/local
+go build && sudo mv hakrevdns /usr/local/bin
 
-if [ -e /usr/bin/hakrevdns ]
+if [ -e /usr/local/bin/hakrevdns ]
 then
   echo -e "\e[31mhakrevdns \e[32m has been successfully installed\e[0m"
-  echo -e "\e[32m Test it out by trying, echo "173.0.84.110" | hakrevdns \e[0m"
+  echo -e "\e[32mTest it out by trying, echo "173.0.84.110" | hakrevdns \e[0m"
 fi
 
 

--- a/main.go
+++ b/main.go
@@ -7,6 +7,7 @@ import (
 	"net"
 	"os"
 	"sync"
+	"strings"
 
 	flags "github.com/jessevdk/go-flags"
 )
@@ -15,6 +16,7 @@ var opts struct {
 	ResolverIP string `short:"r" long:"resolver" description:"IP of the DNS resolver to use for lookups"`
 	Protocol   string `short:"P" long:"protocol" choice:"tcp" choice:"udp" default:"udp" description:"Protocol to use for lookups"`
 	Port       uint16 `short:"p" long:"port" default:"53" description:"Port to bother the specified DNS resolver on"`
+	Domain 	   bool   `short:"d" long:"domain" description:"Output only domains"`
 }
 
 func worker(ip string, wg *sync.WaitGroup, res chan string) {
@@ -38,7 +40,11 @@ func worker(ip string, wg *sync.WaitGroup, res chan string) {
 	}
 
 	for _, a := range addr {
-		res <- fmt.Sprintf("%s \t %s", ip, a)
+		if opts.Domain {
+			res <- fmt.Sprintf("%s", strings.TrimRight(a, "."))
+		} else {
+			res <- fmt.Sprintf("%s \t %s", ip, a)
+		}
 	}
 }
 

--- a/main.go
+++ b/main.go
@@ -1,77 +1,77 @@
 package main
 
 import (
-	"bufio"
-	"context"
-	"fmt"
-	"net"
-	"os"
-	"sync"
-	"strings"
-
-	flags "github.com/jessevdk/go-flags"
+    "bufio"
+    "context"
+    flags "github.com/jessevdk/go-flags"
+    "fmt"
+    "net"
+    "sync"
+    "strings"
+    "os"
 )
 
 var opts struct {
-	ResolverIP string `short:"r" long:"resolver" description:"IP of the DNS resolver to use for lookups"`
-	Protocol   string `short:"P" long:"protocol" choice:"tcp" choice:"udp" default:"udp" description:"Protocol to use for lookups"`
-	Port       uint16 `short:"p" long:"port" default:"53" description:"Port to bother the specified DNS resolver on"`
-	Domain 	   bool   `short:"d" long:"domain" description:"Output only domains"`
-}
-
-func worker(ip string, wg *sync.WaitGroup, res chan string) {
-	defer wg.Done()
-
-	var r *net.Resolver
-
-	if opts.ResolverIP != "" {
-		r = &net.Resolver{
-			PreferGo: true,
-			Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
-				d := net.Dialer{}
-				return d.DialContext(ctx, opts.Protocol, fmt.Sprintf("%s:%d", opts.ResolverIP, opts.Port))
-			},
-		}
-	}
-
-	addr, err := r.LookupAddr(context.Background(), ip)
-	if err != nil {
-		return
-	}
-
-	for _, a := range addr {
-		if opts.Domain {
-			res <- fmt.Sprintf("%s", strings.TrimRight(a, "."))
-		} else {
-			res <- fmt.Sprintf("%s \t %s", ip, a)
-		}
-	}
+        Threads int `short:"t" long:"threads" default:"8" description:"How many threads should be used"`
+        ResolverIP string `short:"r" long:"resolver" description:"IP of the DNS resolver to use for lookups"`
+        Protocol   string `short:"P" long:"protocol" choice:"tcp" choice:"udp" default:"udp" description:"Protocol to use for lookups"`
+        Port       uint16 `short:"p" long:"port" default:"53" description:"Port to bother the specified DNS resolver on"`
+	Domain     bool   `short:"d" long:"domain" description:"Output only domains"`
 }
 
 func main() {
-	_, err := flags.ParseArgs(&opts, os.Args)
-	if err != nil {
-		os.Exit(1)
-	}
+        _, err := flags.ParseArgs(&opts, os.Args)
+        if err != nil{
+            os.Exit(1)
+        }
 
-	var wg sync.WaitGroup
-	res := make(chan string)
+        // default of 8 threads
+        numWorkers := opts.Threads
 
-	scanner := bufio.NewScanner(os.Stdin)
-	for scanner.Scan() {
-		wg.Add(1)
-		go worker(scanner.Text(), &wg, res)
-	}
-	if err := scanner.Err(); err != nil {
-		fmt.Fprintln(os.Stderr, err)
-	}
+        work := make(chan string)
+        go func() {
+            s := bufio.NewScanner(os.Stdin)
+            for s.Scan() {
+                work <- s.Text()
+            }
+            close(work)
+        }()
 
-	go func() {
-		wg.Wait()
-		close(res)
-	}()
+        wg := &sync.WaitGroup{}
 
-	for r := range res {
-		fmt.Println(r)
-	}
+        for i := 0; i < numWorkers; i++ {
+            wg.Add(1)
+            go doWork(work, wg)
+        }
+        wg.Wait()
+}
+
+func doWork(work chan string, wg *sync.WaitGroup) {
+    defer wg.Done()
+    var r *net.Resolver
+
+    if opts.ResolverIP != "" {
+            r = &net.Resolver{
+                    PreferGo: true,
+                    Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+                            d := net.Dialer{}
+                            return d.DialContext(ctx, opts.Protocol, fmt.Sprintf("%s:%d", opts.ResolverIP, opts.Port))
+                    },
+            }
+    }
+
+    for ip := range work {
+        addr, err := r.LookupAddr(context.Background(), ip)
+        if err != nil {
+                continue
+        }
+
+        for _, a := range addr {
+		if opts.Domain {
+			fmt.Println(strings.TrimRight(a, "."))
+		} else {
+                	fmt.Println(ip, "\t",a)
+		}
+        }
+    }
 }

--- a/main.go
+++ b/main.go
@@ -7,10 +7,7 @@ import (
     "fmt"
     "net"
     "sync"
-<<<<<<< master
     "strings"
-=======
->>>>>>> master
     "os"
 )
 
@@ -19,10 +16,7 @@ var opts struct {
         ResolverIP string `short:"r" long:"resolver" description:"IP of the DNS resolver to use for lookups"`
         Protocol   string `short:"P" long:"protocol" choice:"tcp" choice:"udp" default:"udp" description:"Protocol to use for lookups"`
         Port       uint16 `short:"p" long:"port" default:"53" description:"Port to bother the specified DNS resolver on"`
-<<<<<<< master
 	Domain     bool   `short:"d" long:"domain" description:"Output only domains"`
-=======
->>>>>>> master
 }
 
 func main() {
@@ -73,15 +67,11 @@ func doWork(work chan string, wg *sync.WaitGroup) {
         }
 
         for _, a := range addr {
-<<<<<<< master
 		if opts.Domain {
 			fmt.Println(strings.TrimRight(a, "."))
 		} else {
                 	fmt.Println(ip, "\t",a)
 		}
-=======
-                fmt.Println(ip, "\t",a)
->>>>>>> master
         }
     }
 }

